sled-storage/src/manager.rs

```diff
@@ -764,6 +973,77 @@ impl StorageManager {
         }
     }
 
+    // Ensures a dataset exists within a zpool, according to `config`.
+    async fn ensure_dataset(
+        &mut self,
+        config: &DatasetConfig,
+    ) -> Result<(), Error> {
+        info!(self.log, "ensure_dataset"; "config" => ?config);
+
+        // We can only place datasets within managed disks.
+        // If a disk is attached to this sled, but not a part of the Control
+        // Plane, it is treated as "not found" for dataset placement.
+        if !self
+            .resources
+            .disks()
+            .iter_managed()
+            .any(|(_, disk)| disk.zpool_name() == config.name.pool())
+        {
+            return Err(Error::ZpoolNotFound(format!(
+                "{}",
+                config.name.pool(),
+            )));
+        }
+
+        let zoned = config.name.dataset().zoned();
+        let mountpoint_path = if zoned {
+            Utf8PathBuf::from("/data")
+        } else {
+            config.name.pool().dataset_mountpoint(
+                &Utf8PathBuf::from("/"),
+                &config.name.dataset().to_string(),
+            )
+        };
+        let mountpoint = Mountpoint::Path(mountpoint_path);
+
+        let fs_name = &config.name.full_name();
+        let do_format = true;
+
+        // The "crypt" dataset needs these details, but should already exist
+        // by the time we're creating datasets inside.
+        let encryption_details = None;
+        let size_details = Some(illumos_utils::zfs::SizeDetails {
+            quota: config.quota,
+            reservation: config.reservation,
+            compression: config.compression.clone(),
+        });
+        Zfs::ensure_filesystem(
+            fs_name,
+            mountpoint,
+            zoned,
+            do_format,
+            encryption_details,
+            size_details,
+            None,
+        )?;
+        // Ensure the dataset has a usable UUID.
+        if let Ok(id_str) = Zfs::get_oxide_value(&fs_name, "uuid") {
+            if let Ok(id) = id_str.parse::<DatasetUuid>() {
+                if id != config.id {
```

Comment:
How would this happen? We've asked to set a UUID, which succeeded previously, but somehow it didn't "take"? Related, is there a reason we don't just call `Zfs::set_oxide_value()` unconditionally?

Reply:
I view it as more of a defensive guard against mis-matched configuration, or against "an old device/dataset was imported, and we want to be careful not to auto-import + overwrite it without some explicit intervention".

I don't think this is *likely*, but the name for this dataset is basically something like:

`oxp_<pool UUID>/crucible`

This name is **not that unique**, and doesn't really give a unique identifier that signifies **this** control plane is managing the disk.

By adding (and checking!) the UUID here, we ensure not only "this is a crucible dataset", but also, "this is **our** crucible dataset, and not one configured by someone else".